src/server/requesters/AzureDevOpsRequester.ts

```diff
@@ -0,0 +1,269 @@
+    });
+};
+
+export class AzureDevOpsRequester implements Requester<AzureDevOpsRequesterConfig, AzDOOTPRequestMetadata>{
+
+    slug: string = 'azuredevops-release';
+
+    getConfigForProject(project: Project): AzureDevOpsRequesterConfig | null {
+        return project.requester_AzureDevOps || null;
+    }
+    async metadataForInitialRequest(req: Request, res: Response): Promise<AzDOOTPRequestMetadata | null> {
+        const result = validateMetadataObject(req.body);
+
+        if (result.error) {
+            res.status(400).json({
+                error: 'Request Validation Error',
+                message: result.error.message,
+            });
+            return null;
+        }
+        return { releaseId: result.value.releaseId };
+    }
+    async validateActiveRequest(request: OTPRequest<AzDOOTPRequestMetadata, unknown>, config: AzureDevOpsRequesterConfig): Promise<AllowedState> {
+        const axios = getAxiosForConfig(config);
+        const response = await axios.get<AzDORelease>(`release/releases/${request.requestMetadata.releaseId}`);
+
+        if (response.status !== 200) {
+            return {
+                ok: false,
+                error: "Release with id " + request.requestMetadata.releaseId + " does not exist!"
+            }
+        }
+
+        if (response.data.operationStatus != 'phaseInProgress') {
+            return {
+                ok: false,
+                error: "Release with id " + request.requestMetadata.releaseId + " is not in progress!"
+            }
+        }
+
+        return {
+            ok: true
+        }
+    }
+    isOTPRequestValidForRequester(request: OTPRequest<unknown, unknown>): Promise<OTPRequest<AzDOOTPRequestMetadata, unknown> | null> {
+        const result = validateMetadataObject(request);
+        return result.error ? null : request as any;
+    }
+
+    async getRequestInformationToPassOn(request: OTPRequest<AzDOOTPRequestMetadata, unknown>): Promise<RequestInformation> {
+        const { project } = request;
+
+        const axios = getAxiosForConfig(project.requester_AzureDevOps!);
+        const release = await axios.get<AzDORelease>(`release/releases/${request.requestMetadata.releaseId}`);
+
+        return {
+            description: `Azure DevOps Release ${project.repoOwner}/${project.repoName}#${request.requestMetadata.releaseId}`,
+            url: release.data._links.web.href
+        }
+    }
+
+    validateProofForRequest(
+        request: OTPRequest<AzDOOTPRequestMetadata, unknown>,
+        config: AzureDevOpsRequesterConfig
+    ): Promise<boolean> {
+
+        async function attemptValidateProof(attempts: number) {
+            if (attempts <= 0) {
+                return false;
+            }
+
+            const again = async () => {
+                await new Promise(r => setTimeout(r, 5000));
+                return attemptValidateProof(attempts - 1);
+            };
+
+            return getLogs(config, request.requestMetadata.releaseId)
```

Comment:
Can this be converted to use `async/await` and `try {} catch {}` instead

Reply:
Sure, we can. 

Do we want to do that for consistency reasons? Or something else?

Reply:
Partly consistency but mostly to ensure we maintain the promise chain for things like top level error handling.  Using .then in some places can cause the chain to become detached meaning if the promise rejects it will be an unhandled rejection.  Using async await everywhere maintains the chain much easier

Reply:
> Using .then in some places can cause the chain to become detached meaning if the promise rejects it will be an unhandled rejection.

Hm, I did not know that. Looking into it. Did you know it a resource I can use to learn about that?